fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/security/KeyManager.java

```diff
@@ -22,46 +26,125 @@
 import java.security.GeneralSecurityException;
 import java.util.Arrays;
 import java.util.Base64;
+import java.util.Comparator;
+import java.util.Optional;
 
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
+/** Service manging keys and providing easy access to the main device key and PoP Tokens */
 @Singleton
 public class KeyManager {
 
   private static final String TAG = KeyManager.class.getSimpleName();
 
   private final AndroidKeysetManager keysetManager;
   private final Wallet wallet;
+  private KeyPair keyPair;
 
   @Inject
   public KeyManager(AndroidKeysetManager keysetManager, Wallet wallet) {
     this.keysetManager = keysetManager;
     this.wallet = wallet;
 
     try {
+      regenerateMainKey();
       Log.d(TAG, "Public Key = " + getMainPublicKey().getEncoded());
     } catch (IOException | GeneralSecurityException e) {
-      Log.e(TAG, "Failed to retrieve public key", e);
-      throw new IllegalStateException("Failed to retrieve public key", e);
+      Log.e(TAG, "Failed to retrieve device's key", e);
+      throw new IllegalStateException("Failed to retrieve device's key", e);
     }
   }
 
-  public PublicKey getMainPublicKey() throws IOException, GeneralSecurityException {
-    return getPublicKey(keysetManager.getKeysetHandle());
+  /**
+   * This will regenerate the cached device KeyPair.
+   *
+   * <p>Use this only if you know what you are doing
+   *
+   * @throws IOException when the key cannot be retrieved due to IO errors
+   * @throws GeneralSecurityException when the retrieved key is not valid
+   */
+  public void regenerateMainKey() throws GeneralSecurityException, IOException {
```

Comment:
Except from the constructor, when/where will this method be called ?

Reply:
and isn't it a problem that we're regenerating at every initialization?

Reply:
It is not really generated. More retrieved from the library.

I will rename the function to `cacheMainKey()` and set it private as there is currently no reason to call it from outside the service.